### DISCUSSION TITLE
Performance configurations?

### DISCUSSION THREAD

**Discussion by decimoseptimo** [roleoroleo/yi-hack-Allwinner-v2#689] [General]
I just upgraded several cameras (Yi 1080p Home QFUS) from 0.2.5 to 0.2.9. I know for a fact that they won't stream/capture anything unless `disable cloud=true`. I also disable snapshots.

This way they work decently during daytime, but they still seem to struggle a little at night (recordings are very choppy in frigate). What other meaninful configs are possible to increase camera performance? 

I looked up the FAQs but didn't find info on this subject.

**Comment by roleoroleo** [comment]
The feature that creates the biggest problems about performance is the snapshot.
Other features are not so bad.
Try to disable snapshot if you don't need it.
What about the swap file? Are you using it?
If you enable snapshot, enable it, otherwise try to disable it.

**Comment by nemonoid** [comment]
I am used to tweak my Linux computers and my Android phones, also because I usually have old/underpowered stuff.

So some time ago I had a look at what tweaks I could apply to my YI Home camera, and here is what I did :

**-- DISCLAIMER: I'm not suggesting you do this, I'm doing this just to experiment, and I'm not responsible if it breaks your devices or anything else --**

I added the following code to the beginning of system.sh  (altough I hear it's better to put it in /tmp/sd/yi-hack/startup.sh) :

- This prevents the kernel from making a memory dump of the offending processes if it runs out of ram, which would further bog down the device:
`sysctl -w vm.oom_dump_tasks=0`

- By default YI sets vm.vfs_cache_pressure very high (I think 400 or something) which could cause increased CPU usage trying to find more cached stuff to free. I tried setting this to the usual of 100 instead. 
`sysctl -w vm.vfs_cache_pressure=100`

- With this I try to disable ASLR (Warning: this can compromise IT security !!) because it might increase memory fragmentation with low-ram devices.
`sysctl -w kernel.randomize_va_space=0`

- Here i switched off I/O statistics collection on the SD card:
`echo 0 > /sys/block/mmcblk0/queue/iostats`

- And here I tried to optimize the tunables of the CFQ I/O scheduler. This is taken from an Android I/O optimization guide.
```
echo 4 > /sys/block/mmcblk0/queue/iosched/quantum
echo 80 > /sys/block/mmcblk0/queue/iosched/fifo_expire_sync
echo 330 > /sys/block/mmcblk0/queue/iosched/fifo_expire_async
echo 12582912 > /sys/block/mmcblk0/queue/iosched/back_seek_max
echo 1 > /sys/block/mmcblk0/queue/iosched/back_seek_penalty
echo 60 > /sys/block/mmcblk0/queue/iosched/slice_sync
echo 50 > /sys/block/mmcblk0/queue/iosched/slice_async
echo 2 > /sys/block/mmcblk0/queue/iosched/slice_async_rq
echo 0 > /sys/block/mmcblk0/queue/iosched/slice_idle
echo 0 > /sys/block/mmcblk0/queue/iosched/group_idle
echo 1 > /sys/block/mmcblk0/queue/iosched/low_latency
echo 300 > /sys/block/mmcblk0/queue/iosched/target_latency
```


- I mounted the SD card with the noatime option (I added this at the end of system.sh)
```
mount -o remount,noatime /tmp/sd
mount -o remount,noatime /tmp/sd/yi-hack/www/record/

```

For the moment I don't have problems to report, and the read/write speed on my SD card has slightly increased (tested with dd), but overall no significant performance gains.

**Comment by roleoroleo** [comment]
@nemonoid 
Thank you for this contribution.
I will try your changes asap.

**Reply by nemonoid** [reply]
- The **sysctl** variables are all well documented (they are standard Linux kernel variables), just google them and you will find detailed info..


- Regarding the I/O Scheduler, the YI Cam uses CFQ by default (CFQ is good for throughput and heavy I/O load), but you could also use NOOP (which uses less CPU but is not good if there are a lot of simultaneous read and write operations), or DEADLINE (good for responsiveness, less good for heavy I/O load).

I took the tweaked CFQ variables from this very interesting thread (from Android custom ROM developers):
https://groups.google.com/g/francosdevteam/c/X-Eu1PVI0V8

- Mounting a partition with the **noatime** option is a well established tweak in Linux, you can find a lot of info on it.  Just note that you have to run those 2 commands at the end (or after) system.sh , because otherwise system.sh re-mounts them without **noatime**.

Actually the YI Cam already mounts partitions with the **relatime** option, as all modern Linux kernels do, but **noatime** pushes the gains even further.

**Comment by decimoseptimo** [comment]
@roleoroleo 
Yes swap file is enabled by default.

It seems there isn't much room for improvement, and I wouldn't like to touch the system for no significant performance gains.

**Comment by nemonoid** [comment]
As a final note, on my **Yi 1080p Home RFUS y291ga** the newest firmware **12.0.35*** constantly uses about 10% more CPU than the older **9.0.35*** firmware.

Maybe it's because they included more processing for their new AI recognition features.

(you can check how much CPU is being used by connecting with SSH and issuing the "**top**" command)

Also, if you ever have a doubt if the problem you are having is because the CPU isn't keeping up, you can check "top" **while** the camera is doing its things...  if you don't see the CPU usage reach 100% it should not be a CPU bottleneck. 
(top shows the cpu usage divided by user, system, etc.  idle is the % of cpu that is not being used).

Personally I've never seen the CPU usage going above 78%(user+system) with firmware 9.0.35 and 86%(user+system) with 12.0.35 firmware...  But I only use FTP push and one RSTP server, all the rest is disabled on my cam.

But yes, the CPU is just a very weak ARM single core, I'm already surprised it manages to do movement detection !

Also the situation with the RAM isn't great, just about 11MB free (of the 64MB total) at all times, for sure there is not a lot of wiggle room!  Best thing is to disable the services that one doesn't use.

**Comment by roleoroleo** [comment]
I added your changes to my cam and tested for a week.
It seems working properly so I will add the canghes to the master.
Thank you.
